tools/integration_tests_runner.py

```diff
@@ -1,167 +1,232 @@
+
+    tmpdir = path.devnull
+    out, err, result = None, None, None
+    try:
+        if test_file in TEMPFOLDER_TESTS:
+            tmpdir = tempfile.mkdtemp()
+            params[DATA_PATH_KEY] = tmpdir
 
-class IntegrationRunner:
-    def __init__(self):
-        self.process_cli()
-
-        self.proc_count = multiprocessing.cpu_count()
-        logging.info("Number of processors is: {nproc}".format(nproc=self.proc_count))
-
-        self.file_lock = Lock()
-        self.start_finish_lock = Lock()
-        self.tests = Queue()
-        self.key_postfix = ""
-        if self.user_resource_path:
-            self.key_postfix += ' --user_resource_path="{0}"'.format(self.user_resource_path)
-        if self.data_path:
-            self.key_postfix += ' --data_path="{0}"'.format(self.data_path)
-
-
-    def run_tests(self):
-        intermediate_tests = []
-        for exec_file in self.runlist:
-            intermediate_tests.extend(map(lambda x: (exec_file, x), self.get_tests_from_exec_file(exec_file, "--list_tests")[0]))
-
-        shuffle(intermediate_tests)
-        for test in intermediate_tests:
-            self.tests.put(test)
-
-        with open(self.output, "w") as self.file, open("start-finish.log", "w") as self.start_finish_log:
-            self.run_parallel_tests()
-
-
-    def run_parallel_tests(self):
-        threads = []
-
-        for i in range(0, self.proc_count):
-            thread = Thread(target=self.exec_tests_in_queue)
-            thread.start()
-            threads.append(thread)
+        keys = params_from_dict(params)
+        out, err, result = execute_file(test_file, keys)
 
-        for thread in threads:
-            thread.join()
+    except:
+        logging.exception("Failed to remove the temporary folder {folder}".format(folder=tmpdir))
 
+    finally:
+        if test_file in TEMPFOLDER_TESTS:
+            shutil.rmtree(tmpdir, ignore_errors=True)
 
-    def exec_tests_in_queue(self):
-        while not self.tests.empty():
-            try:
-                test_file, test = self.tests.get()
-                self.exec_test(test_file, test, clean_env=(test_file in TEMPFOLDER_TESTS))
+    return (test_file, test_name), (err, result)
 
-            except:
-                logging.error(traceback.format_exc())
-                return
 
+def params_from_dict(params_dict):
+    return " ".join(
+        '"{0}={1}"'.format(key, value)
+        for key, value in params_dict.iteritems()
+        if value
+    )
 
-    def log_start_finish(self, test_file, keys, start=False, finish=False):
-        if not self.write_start_finish_log:
-            return
 
-        if (not start and not finish) or (start and finish):
-            logging.warning("You need to pass either start=True or finish=True, but only one of them! You passed start={0}, finish={1}".format(start, finish))
-            return
+class IntegrationRunner:
+    def __init__(self):
+        proc_count = cpu_count() / 5 * 4
```

Comment:
I am trying to leave some of the cores free from the tasks spawned by our script, so that the system has cores to run on.